lib/upipe-modules/upipe_audio_merge.c

```diff
@@ -0,0 +1,618 @@
+    struct ubuf_mgr *ubuf_mgr;
+    /** flow format packet */
+    struct uref *flow_format;
+    /** ubuf manager request */
+    struct urequest ubuf_mgr_request;
+
+    /** list of input subpipes */
+    struct uchain inputs;
+
+    /** output pipe */
+    struct upipe *output;
+    /** flow_definition packet */
+    struct uref *flow_def;
+    /** output state */
+    enum upipe_helper_output_state output_state;
+    /** list of output requests */
+    struct uchain request_list;
+
+    /** flow def of the first input subpipe, used to check if the format
+     *  of subsequent subpipes match */
+    struct uref *sub_flow_def;
+
+    /** manager to create input subpipes */
+    struct upipe_mgr sub_mgr;
+
+    /** upump manager */
+    struct upump_mgr *upump_mgr;
+    /** watcher */
+    struct upump *upump;
+
+    /** channels **/
+    uint8_t channels;
+
+    /** public upipe structure */
+    struct upipe upipe;
+};
+
+UPIPE_HELPER_UPIPE(upipe_audio_merge, upipe, UPIPE_AUDIO_MERGE_SIGNATURE)
+UPIPE_HELPER_UREFCOUNT(upipe_audio_merge, urefcount, upipe_audio_merge_no_input)
+UPIPE_HELPER_OUTPUT(upipe_audio_merge, output, flow_def, output_state, request_list)
+UPIPE_HELPER_FLOW(upipe_audio_merge, "sound.")
+UPIPE_HELPER_UBUF_MGR(upipe_audio_merge, ubuf_mgr, flow_format,
+                      ubuf_mgr_request,
+                      upipe_audio_merge_check,
+                      upipe_audio_merge_register_output_request,
+                      upipe_audio_merge_unregister_output_request)
+
+UPIPE_HELPER_UPUMP_MGR(upipe_audio_merge, upump_mgr);
+UPIPE_HELPER_UPUMP(upipe_audio_merge, upump, upump_mgr);
+
+UBASE_FROM_TO(upipe_audio_merge, urefcount, urefcount_real, urefcount_real)
+
+/** @hidden */
+static void upipe_audio_merge_free(struct urefcount *urefcount_real);
+
+/** @internal @This is the private context of an output of an audio_merge
+ * pipe. */
+struct upipe_audio_merge_sub {
+    /** refcount management structure */
+    struct urefcount urefcount;
+    /** structure for double-linked lists */
+    struct uchain uchain;
+
+    /** current input flow definition packet */
+    struct uref *flow_def;
+
+    /** the single uref of incoming audio */
+    struct uref *uref;
+
+    /** public upipe structure */
+    struct upipe upipe;
+};
+
+UPIPE_HELPER_UPIPE(upipe_audio_merge_sub, upipe,
+                   UPIPE_AUDIO_MERGE_INPUT_SIGNATURE)
+UPIPE_HELPER_UREFCOUNT(upipe_audio_merge_sub, urefcount,
+                       upipe_audio_merge_sub_free)
+UPIPE_HELPER_VOID(upipe_audio_merge_sub)
+
+UPIPE_HELPER_SUBPIPE(upipe_audio_merge, upipe_audio_merge_sub, input,
+                     sub_mgr, inputs, uchain)
+
+/** @internal @Checks if two flow defs match, except for channel count
+ *
+ * @param one first uref
+ * @param two second uref
+ * @return true if flow defs match, false if not
+ */
+static bool upipe_audio_merge_match_flowdefs(struct uref *one, struct uref *two)
+{
+    return uref_sound_flow_cmp_rate(one, two) == 0
```

Comment:
ubase_check for all of these

Reply:
Not for the _cmp_ functions `return 0 if both attributes are absent or identical`